File: src/main/java/com/logosprog/mainutils/gui/animation/SwingTimerAnimation.java
/*
 * Copyright (c) 2015. This code is a LogosProg property. All Rights Reserved.
 */

package com.logosprog.mainutils.gui.animation;

import javax.swing.*;
import java.awt.event.ActionListener;

/**
 * Created by forando on 23.06.15.<br>
 * Performs GUI com.logosprog.display.animation on a basis of {@link Timer} events.
 */
public abstract class SwingTimerAnimation<T> {
    protected Timer timer;

    /**
     * The object (normally GUI object) on which com.logosprog.display.animation is performed.
     */
    protected T mainObject;

    protected AnimationListener<T> listener;

    /**
     * <b>Attention!</b> {@link #createTimer(int, ActionListener)} must be called right
     * after this constructor invocation.
     * @param obj The object (normally GUI object) on which com.logosprog.display.animation is performed.
     */
    protected SwingTimerAnimation(T obj){
        this.mainObject = obj;
    }

    /**
     * Initialize the {@link Timer} object.
     * @param interval Interval at which the timer <b>actionPerformed</b> events will be fired up.
     * @param actionListener The listener interface for receiving <b>actionPerformed</b> events
     *                       and providing some com.logosprog.display.animation with {@link #mainObject}.
     */
    protected void createTimer(int interval, ActionListener actionListener){
        timer = new Timer(interval, actionListener);
    }

    /**
     * Starts the timer with predefined settings.
     */
    public void start(){
        if (listener != null) listener.onAnimationStart(mainObject);
        timer.start();
    }

    /**
     * Stops the timer.
     */
    public void stop(){
        if (listener != null) listener.onAnimationEnd(mainObject);
        timer.stop();
    }

    public void addListener(AnimationListener<T> listener) {
        this.listener = listener;
    }
}

File: src/main/kotlin/com/logosprog/mainutils/gui/animation/SwingTimerAnimation.kt
/*
 * Copyright (c) 2016. This code is a LogosProg property. All Rights Reserved.
 */

/*
 * Copyright (c) 2016. This code is a LogosProg property. All Rights Reserved.
 */

@file:Suppress("unused", "MemberVisibilityCanPrivate")

package com.logosprog.mainutils.gui.animation

import java.awt.event.ActionListener
import javax.swing.Timer

/**
 * Performs GUI com.logosprog.display.animation on a basis of {@link Timer} events.<br>
 * **Attention!** [.createTimer] must be called right
 * after this constructor invocation.
 * @param mainObject The object (normally GUI object) on which com.logosprog.display.animation is performed.
 */
abstract class SwingTimerAnimation<T> protected constructor(protected val mainObject: T) {
    protected var timer: Timer? = null

    protected var listener: AnimationListener<T>? = null

    /**
     * Initialize the [Timer] object.
     * @param interval Interval at which the timer **actionPerformed** events will be fired up.
     * *
     * @param actionListener The listener interface for receiving **actionPerformed** events
     * *                       and providing some com.logosprog.display.animation with [.mainObject].
     */
    protected fun createTimer(interval: Int = 1000, actionListener: ActionListener) {
        timer = Timer(interval, actionListener)
    }

    /**
     * Starts the timer with predefined settings.
     */
    open fun start() {
        listener!!.onAnimationStart(mainObject)
        timer?.start()
    }

    /**
     * Stops the timer.
     */
    open fun stop() {
        listener!!.onAnimationEnd(mainObject)
        timer?.stop()
    }

    fun addListener(listener: AnimationListener<T>) {
        this.listener = listener
    }
}
